### PR TITLE
Store & retrieve job as hash instead of jid (WIP, DO NOT MERGE)

### PR DESCRIPTION
**Pull request by Darep**
_Work in progress, needs more testing, need to get rid of `eval`._

It looks to me like the output of `yield` in Sidekiq client middleware's API was changed at some point: it doesn't return a job ID in my app, but instead a hash which contains the job id. Thus, the `store_expiry` method saved a string representation of the hash, and the plugin never found the previously scheduled job and never debounced anything.

I changed the code to adjust to the `yield` change and now it works beautifully! All the current tests were passing even before this change, so I think we should try to come up with a test that first fails and then add this to fix that test.

PS. The `yield` change is not mentioned anywhere in the changelogs, so I'm a bit sceptical that the reason might be e.g. my application config or something else.

### CODE REVIEW

File: lib/sidekiq/debounce.rb [hunk 1]
```diff
@@ -11,20 +11,25 @@ def call(worker, msg, _queue, redis_pool)
...
+        return false if scheduled_job
+        job
       end
     end
 
     private
 
+    def retrieve_scheduled(conn)
+      job_json = conn.get(debounce_key)
+      job_json ? eval(job_json) : nil
```
Comment (Darep) [inline lib/sidekiq/debounce.rb:30]: This is true and I don't really want to use eval. Need to fix this somehow.

**Comment by NuckChorris** [comment]
I've run into similar edgecases during development but they just seemed to go away and I was never able to pin them down fully so I assumed they resolved themselves with other changes :/

Sidekiq's interfaces are almost entirely undocumented so I can't unit test anything, and they're not reliable either so it's hard to write integration tests.

As best as I can tell, in certain circumstances it gives a whole object instead of just the JID, but I have no idea why

**Comment by Darep** [comment]
Ok, that's kinda weird, but good hear. I'll just have to investigate then :smiling_imp:

**Comment by NuckChorris** [comment]
Yeah I think I remember rubber ducking through part of the Sidekiq codebase and finding nothing, so I may have to give it another look

**Comment by borc** [comment]
AFAICT, the job hash is only ever used for retrieving the job ID. Is it really necessary to do all this work, when you could just grab the job ID from the hash returned by `yield`?

Am I missing something?

**Comment by Darep** [comment]
This change makes it so that yield can return whatever it wants and this still works :) Some Sidekiq versions return a job id, some versions return a hash, and who knows what future versions will return ;)

**Comment by borc** [comment]
Ah yes, that. 

But doesn't that mean `reschedule` is now broken for sidekiq versions below 3? First `yield` would return a jid, and that would get stored in the debounce key. Then on subsequent runs, `reschedule` would attempt to call `['jid']` on the jid itself.

**Comment by Darep** [comment]
> But doesn't that mean reschedule is now broken for sidekiq versions below 3?

Good point, whoops. Back to the drawing board...

**Comment by Darep** [comment]
I'll close this, the solution in #6 is simpler, I'll just add a hash-check around it :)

**Pull request closed without merging**